dkg/pedersen/pedersen.go

```diff
@@ -488,6 +491,230 @@ func (w worker) work(jobIndex int) error {
+	}
+
+	txMsg := types.NewReencryptRequest(U, pubk)
+
+	err = <-sender.Send(txMsg, addrs...)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to send reencrypt request: %v", err)
+	}
+
+	ocs := newOCS(pubk)
+	ocs.U = U
+	ocs.nbnodes = len(addrs)
+	ocs.threshold = threshold
+
+	for i := 0; i < len(addrs); i++ {
+		src, rxMsg, err := receiver.Recv(ctx)
+		if err != nil {
+			return nil, xerrors.Errorf(unexpectedStreamStop, err)
+		}
+
+		dela.Logger.Debug().Msgf("Received a decryption reply from %v", src)
+
+		reply, ok := rxMsg.(types.ReencryptReply)
+		if !ok {
+			return nil, xerrors.Errorf(unexpectedReply, reply, rxMsg)
+		}
+
+		err = processReencryptReply(ocs, &reply)
+		if err == nil {
+			dela.Logger.Info().Msgf("Reencrypted message: %v", ocs.Uis)
+
+			return ocs.Uis, nil
+		}
+	}
+
+	return nil, xerrors.Errorf("Reencryption failed: %v", err)
+}
+
+func processReencryptReply(ocs *OCS, reply *types.ReencryptReply) (err error) {
+	if reply.Ui == nil {
+		err = xerrors.Errorf("Received empty reply")
+		dela.Logger.Warn().Msg("Empty reply received")
+		ocs.nbfailures++
+		if ocs.nbfailures > ocs.nbnodes-ocs.threshold {
+			err = xerrors.Errorf("couldn't get enough shares")
+			dela.Logger.Warn().Msg(err.Error())
+		}
+		return err
+	}
+
+	ocs.replies = append(ocs.replies, *reply)
+
+	if len(ocs.replies) >= ocs.threshold {
+		ocs.Uis = make([]*share.PubShare, 0, ocs.nbnodes)
+
+		for _, r := range ocs.replies {
+
+			/*
+				// Verify proofs
+				ufi := suite.Point().Mul(r.Fi, suite.Point().Add(ocs.U, ocs.pubk))
+				uiei := suite.Point().Mul(suite.Scalar().Neg(r.Ei), r.Ui.V)
+				uiHat := suite.Point().Add(ufi, uiei)
+
+				gfi := suite.Point().Mul(r.Fi, nil)
+				gxi := ocs.poly.Eval(r.Ui.I).V
+				hiei := suite.Point().Mul(suite.Scalar().Neg(r.Ei), gxi)
+				hiHat := suite.Point().Add(gfi, hiei)
+				hash := sha256.New()
+				r.Ui.V.MarshalTo(hash)
+				uiHat.MarshalTo(hash)
+				hiHat.MarshalTo(hash)
+				e := suite.Scalar().SetBytes(hash.Sum(nil))
+				if e.Equal(r.Ei) {
+
+			*/
+			ocs.Uis = append(ocs.Uis, r.Ui)
+			/*
+				}
+				else
+				{
+					dela.Logger.Warn().Msgf("Received invalid share from node: %v", r.Ui.I)
+					ocs.nbfailures++
+				}
+			*/
+		}
+		dela.Logger.Info().Msg("Reencryption completed")
+		return nil
+	}
+
+	// If we are leaving by here it means that we do not have
+	// enough replies yet. We must eventually trigger a finish()
+	// somehow. It will either happen because we get another
+	// reply, and now we have enough, or because we get enough
+	// failures and know to give up, or because o.timeout triggers
+	// and calls finish(false) in its callback function.
+
+	err = xerrors.Errorf("not enough replies")
+	dela.Logger.Debug().Msg(err.Error())
+	return err
+}
+
+// Helper functions
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// DecryptSecret implements dkg.Actor.
+func (a *Actor) DecryptSecret(Cs []kyber.Point, XhatEnc kyber.Point, Sk kyber.Scalar) (msg []byte, err error) {
```

Comment:
Conceptually, shouldn't this logic be somewhere else ? 
If I understand correctly, this should be called by a client, normally...

Reply:
I think we misunderstood each other 😅 

I didn't mean the OCS (although it's probably a good thing that you have split it out).

I meant the secret decryption seems off here (as we discussed in person yesterday), and it probably makes more sense into its own library.